testbed/tests/widgets/test_slider.py

```diff
@@ -1,81 +1,221 @@
+from math import pi
 from unittest.mock import Mock
 
-from pytest import approx, fixture, mark
+import pytest
+from pytest import approx, fixture
 
 import toga
-from toga.platform import current_platform
 
-POSITIONS = [0, 0.01, 0.1, 0.5, 0.9, 0.99, 1]
-SCALES = [0.01, 0.1, 1, 10, 100000]
+from ..assertions import assert_set_get
+from .properties import (  # noqa: F401
+    test_enabled,
+    test_flex_horizontal_widget_size,
+)
 
-# How accurate the position must be in continuous mode.
-ACCURACY = 0.001
+# To ensure less than 1 pixel of error, the slider must be able to distinguish at least
+# 10,000 positions in continuous mode.
+#
+# Pi is irrational, which helps test things which should be accurate to within the limits
+# of a Python float, e.g. setting a value and then immediately getting it.
+ACCURACY = 0.0001
+POSITIONS = [0, 0.0001, 1 / pi, 0.5, 0.9, 0.9999, 1]
+SCALES = [0.0001, 0.1, 1, pi, 10000]
 
 
 @fixture
-async def widget(on_change):
-    return toga.Slider(on_change=on_change)
+async def widget():
+    return toga.Slider()
 
 
 @fixture
-def on_change():
-    return Mock()
+def on_change(widget):
+    handler = Mock()
+    widget.on_change = handler
+    return handler
 
 
-@mark.skipif(current_platform == "android", reason="value is 0.0")
-async def test_init(widget, probe, on_change):
+async def test_init(widget, probe):
     assert widget.value == 0.5
     assert widget.range == (0, 1)
-    assert probe.position == 0.5
-    on_change.assert_not_called()
+    assert widget.tick_count is None
+    assert probe.position == approx(0.5, abs=ACCURACY)
 
 
-@mark.skipif(
-    current_platform in ["windows", "macOS", "iOS", "linux"],
-    reason="on_change called 2 times",
-)
-@mark.skipif(current_platform == "android", reason="position is 0.0")
+async def test_init_handlers():
+    handlers = {
+        name: Mock(name=name) for name in ["on_change", "on_press", "on_release"]
+    }
+    toga.Slider(**handlers)
+    for handler in handlers.values():
+        handler.assert_not_called()
+
+
+# Bounds checks are covered by core tests.
 async def test_value(widget, probe, on_change):
     for scale in SCALES:
         widget.range = (0, scale)
         for position in POSITIONS:
-            on_change.mock_calls.clear()
-            widget.value = position * scale
+            on_change.reset_mock()
+            assert_set_value(widget, position * scale)
```

Comment:
OK, I've restored the redraw in test_button, and added redraws to test_slider in all the methods you mentioned below. However, in this case it isn't necessary to redraw between the action and the assertion, and I think it's actually a stronger test if you don't do that, because it verifies that the changed values are available immediately.

Reply:
Ok - that makes sense. The redraw a couple of lines later is enough to do a visual inspection that the value has taken effect.